### PR TITLE
Add (untested) slow path and make performance test compile

### PR DESCRIPTION
**Pull request by solimantwali**
- [x] Rename utilities to utility_lib
- [ ] Debug performance tests

### CODE REVIEW

**Review by thedavidchu — CHANGES_REQUESTED** [review]
I didn't check the algorithm, I mostly checked for code hygiene.

Might as well merge it now with a broken algorithm so that we can run our perf tests sooner rather than later #️⃣approximatecomputing

File: src/parallel/include/parallel/parallel.hpp [hunk 1]
```diff
@@ -86,11 +87,6 @@ struct KeyValue {
 
 using ParallelBucket = std::atomic<KeyValue>;
 
-/// This ensures we are guaranteed to use atomic operations (instead of locks).
-/// This for performance (to follow the algorithm's 'fast-path') rather than for
-/// correctness.
-static_assert(ParallelBucket::is_always_lock_free);
-
```
[the remark this answers is not in the document]
Comment (solimantwali) [reply src/parallel/include/parallel/parallel.hpp]: addressed in 12f40cc7c8b8c8ec2a084cf9720f90fd56fda4da

File: src/parallel/parallel.cpp [hunk 2]
```diff
@@ -154,33 +128,145 @@ bool
...
+    //         break;
+    //     }
+    //     while(curr_kv.key == key) {
+    //         bool updated = this->compare_and_set_key_val(idx, curr_kv, new_kv);
+    //         if (updated == true) {
+    //             return true;
+    //         }
+    //     }
+    // }
+
```
[the remark this answers is not in the document]
Comment (solimantwali) [reply src/parallel/parallel.cpp]: addressed in 2f697247630c8105f8ea0a80c364eb35f4350805

File: src/parallel/parallel.cpp [hunk 3]
```diff
@@ -154,33 +128,145 @@ bool
...
+    size_t curr_index = index;
+    manager.lock(next_index); 
+
+    size_t next_offset = next_index - get_home(hash(atomic_load_key_val(next_index).key), this->capacity_);
+    KeyValue empty_entry = {.key=bucket_empty_key}; 
+
+    // shift to the left
+    while(next_offset > 0) {
+        manager.lock(next_index); 
+        KeyValue entry_to_move = this->do_atomic_swap(empty_entry, next_index); //na this is fucked up
```
[the remark this answers is not in the document]
Comment (solimantwali) [reply src/parallel/parallel.cpp:225]: sad

File: src/parallel/parallel.cpp [hunk 4]
```diff
@@ -201,36 +287,27 @@ ParallelRobinHoodHashTable::remove_thread_lock_manager()
...
 std::pair<size_t, bool>
 ParallelRobinHoodHashTable::find_next_index_lock(ThreadManager &manager,
                                                  size_t start_index,
-                                                 KeyType key,
-                                                 size_t &distance_key)
+                                                 KeyType key)
 {
     LOG_TRACE("Enter");
     const size_t capacity = this->buckets_.size();
     for (size_t i = 0; i < capacity; ++i) {
```
[the remark this answers is not in the document]
Comment (solimantwali) [reply src/parallel/parallel.cpp]: addressed in 42e6bb1481d15db6a27296ebae6bd08eebf915c9

File: test/performance_test/main.cpp [hunk 5]
```diff
@@ -2,6 +2,7 @@
 #include <iostream>
 #include <thread>
 #include <vector>
+#include <functional>
```
[the remark this answers is not in the document]
Comment (solimantwali) [reply test/performance_test/main.cpp]: addressed in 42e6bb1481d15db6a27296ebae6bd08eebf915c9

File: src/parallel/parallel.cpp [hunk 6]
```diff
@@ -154,33 +128,145 @@ bool
...
+    }
+
+    size_t next_index = index++;
+    size_t curr_index = index;
+    manager.lock(next_index); 
+
+    size_t next_offset = next_index - get_home(hash(atomic_load_key_val(next_index).key), this->capacity_);
+    KeyValue empty_entry = {.key=bucket_empty_key}; 
+
+    // shift to the left
```
[the remark this answers is not in the document]
Comment (solimantwali) [reply src/parallel/parallel.cpp]: addressed in 42e6bb1481d15db6a27296ebae6bd08eebf915c9

**Review by thedavidchu — APPROVED** [review]
lgtm (on the condition it compiles with the changes!)

**Pull request merged**